pages/docs/manual/latest/polymorphic-variant.mdx

```diff
@@ -0,0 +1,273 @@
+
+First off, here are some key features:
+
+- Poly variants are **structurally typed** (in comparison to **nominally typed** variants). They can be used without an explicit type definition.
+- They allow easier JavaScript interop (compile to strings / objects with predictable `NAME` and `VAL` attribute) and don't need explicit runtime conversions, unlike common variants.
+- Due their structural nature, they oftentimes cause tricky type checking errors when types don't match up, which makes them a more advanced feature.
+
+## Basics
+
+This is how you'd define a poly variant type with an exact set of constructors:
+
+```res
+// Note the surrounding square brackets, and # for constructors
+type color = [ #Red | #Green | #Blue ]
+```
+
+Here is how you'd construct a poly variant value:
+
+```res
+// This doesn't actually need any color type definition
+// beforehand
+let myColor = #Red
+```
+
+We can also use poly variant types in annotations without an explicit type definition:
+
+```res
+let render = (color: [#Red | #Green | #Blue]) => {
+  switch(color) {
+    | _ => Js.log("...")
+  }
+}
+
+let color: [#Red] = #Red
+```
+
+### Constructor Names
+
+Poly variant constructor names are less restrictive than in common variants (e.g. they don't need to be capitalized):
+
+```res
+type users = [ #admin | #moderator | #user ]
+
+let admin = #admin
+```
+
+In rare cases (mostly for JS interop reasons), it's also possible to define invalid identifiers, such as hypens or numbers:
+
+```res
+type numbers = [#\"1" | #\"2"]
+let one = #\"1"
+```
+
+### Constructor Arguments
+
+This is equivalent to what we've already learned with common variants:
+
+```res
+type account = [
+  | #Anonymous
+  | #Instagram(string)
+  | #Facebook(string, int)
+]
+
+let acc: account = #Instagram("test")
+```
+
+### Annotations with Upper / Lower Bound Constraints
+
+There's also a way to define an "upper" and "lower" bound constraint for a poly variant type (that's why they are called _Polymorphic Variants_). Here is what it looks like in type annotations:
+
+```res
+// Only #Red allowed, no upper / lower bound (= exact)
+let basic: [#Red] = #Red
+
+// May contain #Red, or any other value (open variant)
+// here, foreground will be an inferred type [> #Red | #Green]
+let foreground: [> #Red] = #Green
+
+// The value must be "one of" #Red | #Blue
+// Only #Red and #Blue are valid values
+let background: [< #Red | #Blue] = #Red
+```
+
+Don't worry about the upper / lower bound feature (aka polymorphism) just yet, since this is a very advanced topic that's often not really needed. For the sake of completeness, we mention a few details about it [later on](#lower--upper-bound-constraints).
+
+
+## Polymorphic Variants are Structurally Typed
+
+As we've already seen in the section above, poly variants are treated a little bit differently than common variants. Most notably, we don't need any explicit type definition to define a value.
+
+```res
```

Comment:
You may mention one virtue of structural types, for package a and b who share the same structural types, they don't need depend on each other. But it is not the case for norminal types

Reply:
mentioned it in multiple occasions